### PR TITLE
[purs ide] improve import parsing

### PR DESCRIPTION
**Pull request by kRITZCREEK**
Still needs a few comments, but this replaces the custom parser for adding imports with the real import parser.

### CODE REVIEW

**Comment by paf31** [comment]
Can this reuse the existing parser code?

**Comment by kRITZCREEK** [comment]
It basically does, except for me needing to extract specific position information irrelevant to the existing parser code. The `parseModuleHeader` I wrote could basically be used to speed up the building of the dependency graph, but it's not the "real" solution since it still completely lexes the file in question before starting to parse.

**Comment by paf31** [comment]
But there is still some code which can be shared, right? You could move most of `parseModuleHeader` into the parser and pull the common code out of the actual module parser.

**Comment by kRITZCREEK** [comment]
Like this?

**Comment by paf31** [comment]
Wonderful, thanks.

**Comment by kRITZCREEK** [comment]
Okay I added a few comments and tests. I'll try this out on a few test files as soon as I get the scripts set up properly... The tests are fairly comprehensive, so I'd still say this is good to go.

File: src/Language/PureScript/Parser/Lexer.hs [hunk 1]
```diff
@@ -173,6 +174,12 @@ updatePositions (x:xs) = x : zipWith update (x:xs) xs
 parseTokens :: Lexer u [PositionedToken]
 parseTokens = whitespace *> P.many parsePositionedToken <* P.skipMany parseComment <* P.eof
 
+lexLenient :: FilePath -> Text -> Either P.ParseError [PositionedToken]
+lexLenient f s = updatePositions <$> P.parse parseTokensLenient f s
+
+parseTokensLenient :: Lexer u [PositionedToken]
```
Comment (paf31) [inline src/Language/PureScript/Parser/Lexer.hs]: Could you please add a comment here?

Comment (paf31) [inline src/Language/PureScript/Parser/Lexer.hs]: I assume this is here to handle cases where something wouldn't fully parse normally? If so, could you also please add a test for that?

File: src/Language/PureScript/Parser/Lexer.hs [hunk 2]
```diff
@@ -174,6 +174,8 @@ updatePositions (x:xs) = x : zipWith update (x:xs) xs
 parseTokens :: Lexer u [PositionedToken]
 parseTokens = whitespace *> P.many parsePositionedToken <* P.skipMany parseComment <* P.eof
 
+-- | Lexes the given file, and on encountering a parse error, returns the
```
Comment (paf31) [inline src/Language/PureScript/Parser/Lexer.hs]: Thanks!

File: src/Language/PureScript/Parser/Lexer.hs (hunk 1 above)
Comment (kRITZCREEK) [inline src/Language/PureScript/Parser/Lexer.hs]: Where should I put unit tests for compiler code? I can add them to purs ide's suite, which kind of steals the usual namespace for hspec test suites.

Comment (paf31) [inline src/Language/PureScript/Parser/Lexer.hs]: I just meant that you could maybe add a copy of an existing `psc-ide` test which has a syntax error after the imports somewhere. Does that work?

**Comment by paf31** [comment]
Thanks @kRITZCREEK, looks good!

**Pull request merged**